st2common/st2common/services/action.py

```diff
@@ -154,9 +154,35 @@ def update_status(liveaction, new_status, publish=True):
     return liveaction
 
 
-def is_action_canceled(liveaction_id):
+def is_action_canceled_or_canceling(liveaction_id):
     liveaction_db = action_utils.get_liveaction_by_id(liveaction_id)
-    return liveaction_db.status == action_constants.LIVEACTION_STATUS_CANCELED
+    return liveaction_db.status in [action_constants.LIVEACTION_STATUS_CANCELED,
+                                    action_constants.LIVEACTION_STATUS_CANCELING]
+
+
+def request_cancellation(liveaction, requester):
+    """
+    Request cancellation of an action execution.
+
+    :return: (liveaction, execution)
+    :rtype: tuple
+    """
+    if liveaction.status == action_constants.LIVEACTION_STATUS_CANCELING:
+        return liveaction
+
+    if liveaction.status not in action_constants.CANCELABLE_STATES:
+        raise Exception('Unable to cancel execution because it is already in a completed state.')
+
+    result = {
+        'message': 'Action canceled by user.',
+        'user': requester
+    }
+
+    update_status(liveaction, action_constants.LIVEACTION_STATUS_CANCELING, result=result)
```

Comment:
The end_timestamp is updated in _do_cancel in st2actions/st2actions/container/base.py when the status is updated to CANCELED.